### PR TITLE
Update Dockerfile to use workspaces setup

### LINKED ISSUE #84: Docker "Service 'calendso' failed to build" due to "Step 4/31 : COPY calendso/apps/web/prisma"
**Environment**
1. Fresh install of Ubuntu 20.04.3 LTS Desktop

<br>

**Install Process**
1. Install `Docker` and `Docker-Compose` via the official Docker website procedure
2. Clone calendso-docker (I have also tried this with `--recursive` at the end as mentioned in https://github.com/calcom/docker/issues/12)
`git clone --recursive https://github.com/calendso/docker.git calendso-docker`
3. Change into the directory
`cd calendso-docker`
4. Rename .env.example to .env and update .env.
`cp .env.example .env`
5. Build and start calendso
`docker-compose up --build`

<br>

**Results**
```
$sudo docker-compose up --build
Creating network "calendso-docker_stack" with the default driver
Creating volume "calendso-docker_database-data" with default driver
Pulling database (postgres:)...
latest: Pulling from library/postgres
5eb5b503b376: Pull complete
daa0467a6c48: Pull complete
7cf625de49ef: Pull complete
bb8afcc973b2: Pull complete
c74bf40d29ee: Pull complete
2ceaf201bb22: Pull complete
1255f255c0eb: Pull complete
12a9879c7aa1: Pull complete
4e80271c8de8: Pull complete
04d0275af286: Pull complete
3d4fe3f39342: Pull complete
60f2aefbd6d9: Pull complete
98ace1022c39: Pull complete
Digest: sha256:156c50d4b6fe6ea4e4645ccdeabf54fedc59a561bfece047cdf4c26a42deab72
Status: Downloaded newer image for postgres:latest
Building calendso
Sending build context to Docker daemon   26.1MB
Step 1/31 : FROM node:14 as deps
14: Pulling from library/node
a834d7c95167: Pull complete 
57b3fa6f1b88: Pull complete 
778df3ecaa0f: Pull complete 
d353c340774e: Pull complete 
6370e0bc373d: Pull complete 
fb61153482cd: Pull complete 
78fb5822e501: Pull complete 
ba3577a691be: Pull complete 
bd38fd0dd57b: Pull complete 
Digest: sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865
Status: Downloaded newer image for node:14
 ---> 9cb3f042a684
Step 2/31 : WORKDIR /calcom
 ---> Running in 3cb224df2841
Removing intermediate container 3cb224df2841
 ---> 34be7496817b
Step 3/31 : COPY calendso/apps/web/package.json calendso/apps/web/yarn.lock ./
 ---> f22c2b1deaa7
Step 4/31 : COPY calendso/apps/web/prisma prisma
COPY failed: file not found in build context or excluded by .dockerignore: stat calendso/apps/web/prisma: file does not exist
ERROR: Service 'calendso' failed to build : Build failed
```

### PR DESCRIPTION
**Pull request by searsaw**
The main Cal.com project has started using Yarn workspaces to
break the project up into separate packages and treat the project
as a monorepo. This required quite a few changes to the Dockerfile
that assumed the previous project setup.

This should fix #84 and #85.

### CODE REVIEW

**Comment by searsaw** [comment]
The work on this branch functions. However, to get it to work, I had to comment out the database migrations in the `start.sh` file. The migrations currently fail because `npx` can't find the `prisma` package. This is because we copy the `node_modules` from the `builder` stage. That stage installs the production dependencies as its last step. `prisma` is a development dependency in this project. Therefore it isn't available.

The simple solution would be to install prisma in the image itself so it would be available. However, I left it commented out for now so the community could make a decision together. It's also possible the `prisma` package could be made a production dependency in the main project, but that would take a PR to that project which I imagine would warrant a conversation to determine if they are okay with that course of action.

**Comment by BahrouniYousra** [comment]
By launching the project with these modifications it posed this problem:
"Error: Cannot find module '/root/.npm/_npx/1/lib/node_modules/prisma/scripts/preinstall-entry.js'"

Any ideas how i can solve it plz knowing i tried to generate prisma: npx prisma generate --schema=calendso/packages/prisma/schema.prisma
also to install its dependencies with npm install prisma@latest prisma without success
Thank you for your answers !

**Comment by leprodude** [comment]
After reading this, I'm quite sure it will solve my problem as well. When trying to install I get:

```
...
=> ERROR [deps 4/5] COPY calendso/apps/web/prisma prisma                                                         0.0s
------
 > [deps 4/5] COPY calendso/apps/web/prisma prisma:
------
failed to compute cache key: "/calendso/apps/web/prisma" not found: not found
ERROR: Service 'calendso' failed to build : Build failed
```

So I'll wait until it's merged and try again!

**Comment by searsaw** [comment]
@leprodude That line was changed in this branch.

**Comment by leprodude** [comment]
Ah I could just pull from this branch instead, even before it's merged to main. Haven't really thought of that lol

**Comment by searsaw** [comment]
@leprodude The first issue there is a Docker Compose error. Looks like it tried to reach your Docker server but couldn't. This could have happened if you're using Docker for Mac and the server was still spinning up. For example, if you had just tried restarting it or turning it on or something. Hence why waiting got rid of that error.

The second error is due to what appears to be you using the old Dockerfile still. The line `COPY calendso/apps/web/prisma prisma` which it is blowing up on [was changed in my branch](https://github.com/calcom/docker/pull/86/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L5). Make sure when running Docker Compose that you telling it to rebuild itself. It looks like it's doing that but who knows? If you want to go the nuclear option, which I'm a big fan of, you can do the following:

```shell
# destroy docker compose environment
docker-compose down
# clean up all docker stuff to get a clean slate
docker system prune --all
# recreate docker compose environment, making sure Cal.com is rebuilt each time
docker-compose up --build
```

**Comment by leprodude** [comment]
@searsaw Yes that's what I figured! Then I realised that on my local machine I haven't started my Docker deamon yet :'D (on the server it's usually running already)

Ok I've done what you suggested and reclaimed 5GB of space, so first of all thanks for that!
But after restarting, I still get that damn error. I cloned the repo from your branch.

Do I have to install anything before, or else have a specific version number of node maybe?

```
Creating network "calendso-docker_stack" with the default driver
Pulling database (postgres:)...
latest: Pulling from library/postgres
5eb5b503b376: Pull complete
daa0467a6c48: Pull complete
7cf625de49ef: Pull complete
bb8afcc973b2: Pull complete
c74bf40d29ee: Pull complete
2ceaf201bb22: Pull complete
1255f255c0eb: Pull complete
12a9879c7aa1: Pull complete
4e80271c8de8: Pull complete
04d0275af286: Pull complete
3d4fe3f39342: Pull complete
60f2aefbd6d9: Pull complete
98ace1022c39: Pull complete
Digest: sha256:156c50d4b6fe6ea4e4645ccdeabf54fedc59a561bfece047cdf4c26a42deab72
Status: Downloaded newer image for postgres:latest
Building calendso
[+] Building 3.7s (8/23)                                                                                                                                                       
 => [internal] load build definition from Dockerfile                                                                                                                      0.0s
 => => transferring dockerfile: 1.41kB                                                                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                                                         0.0s
 => => transferring context: 2B                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/node:14                                                                                                                3.2s
 => [internal] load build context                                                                                                                                         0.4s
 => => transferring context: 7.91MB                                                                                                                                       0.4s
 => CANCELED [deps 1/5] FROM docker.io/library/node:14@sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865                                            0.4s
 => => resolve docker.io/library/node:14@sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865                                                          0.0s
 => => sha256:28874e89b1ebebe12162aed108a8b4b37f9e0326a58830d1fe54418f0167cad0 2.21kB / 2.21kB                                                                            0.0s
 => => sha256:778df3ecaa0fbba90d3a7d88947a4376ebdc7e2fcf8a4b5ce43b3c699faadff6 0B / 4.34MB                                                                                0.4s
 => => sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865 776B / 776B                                                                                0.0s
 => => sha256:9cb3f042a68426bdefb8eba9bc173bdab3121c897397843ffc130b909dd86e4b 7.64kB / 7.64kB                                                                            0.0s
 => => sha256:a834d7c95167a3e129adb00a5ddbaf5d3c035ad748ff7ee1273373d150457820 0B / 45.38MB                                                                               0.4s
 => => sha256:57b3fa6f1b88b95ac6adeafdb618011e672d4c9f5637b92be373276ee7e066dd 0B / 11.30MB                                                                               0.4s
 => CACHED [deps 2/5] WORKDIR /calcom                                                                                                                                     0.0s
 => CACHED [deps 3/5] COPY calendso/apps/web/package.json calendso/apps/web/yarn.lock ./                                                                                  0.0s
 => ERROR [deps 4/5] COPY calendso/apps/web/prisma prisma                                                                                                                 0.0s
------
 > [deps 4/5] COPY calendso/apps/web/prisma prisma:
------
failed to compute cache key: "/calendso/apps/web/prisma" not found: not found
ERROR: Service 'calendso' failed to build : Build failed
```

**Comment by searsaw** [comment]
Okay. So let's make sure you're on the right branch. To get the branch, you'd need to clone my repo or add it as another remote to your local and then switch to it. So something like this.

```shell
git clone git@github.com:calcom/docker.git
cd docker
git remote add searsaw git@github.com:searsaw/docker.git
git fetch --all
git checkout searsaw/fix-workspaces-build-issues
docker-composer up --build
```

This should put you on my branch and then build the Docker Compose environment again.

**Comment by leprodude** [comment]
Ok I was just cloning your git@github.com:searsaw/docker.git and thought that would be enough, but of course that takes the code from the main branch, which doesn't have the file changed yet. Still, having done that I get an error building. Going to investigate it myself now, but just in case it's obvious to you I'll leave it here. Thanks already for holding my hand through this so far lol

```
Building calendso
[+] Building 0.8s (16/29)                                                                                                                                                      
 => [internal] load build definition from Dockerfile                                                                                                                      0.0s
 => => transferring dockerfile: 37B                                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                                         0.0s
 => => transferring context: 2B                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/node:14                                                                                                                0.7s
 => [internal] load build context                                                                                                                                         0.0s
 => => transferring context: 168B                                                                                                                                         0.0s
 => CANCELED [runner  1/14] FROM docker.io/library/node:14@sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865                                        0.0s
 => => resolve docker.io/library/node:14@sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865                                                          0.0s
 => => sha256:28874e89b1ebebe12162aed108a8b4b37f9e0326a58830d1fe54418f0167cad0 2.21kB / 2.21kB                                                                            0.0s
 => => sha256:9cb3f042a68426bdefb8eba9bc173bdab3121c897397843ffc130b909dd86e4b 7.64kB / 7.64kB                                                                            0.0s
 => => sha256:b2c75df8c9706156c38b4f1f678d00e11cb2bfda09fc4ab6e36ec17ac9163865 776B / 776B                                                                                0.0s
 => CACHED [runner  2/14] WORKDIR /calcom                                                                                                                                 0.0s
 => ERROR [builder 3/7] COPY calendso/package.json calendso/yarn.lock calendso/turbo.json ./                                                                              0.0s
 => ERROR [builder 4/7] COPY calendso/apps/web ./apps/web                                                                                                                 0.0s
 => ERROR [builder 5/7] COPY calendso/packages ./packages                                                                                                                 0.0s
 => CACHED [runner  3/14] RUN apt-get update &&     apt-get -y install netcat &&     rm -rf /var/lib/apt/lists/* &&     npm install --global prisma                       0.0s
 => ERROR [runner  4/14] COPY calendso/package.json calendso/yarn.lock calendso/turbo.json ./                                                                             0.0s
 => ERROR [deps 3/8] COPY calendso/package.json calendso/yarn.lock ./                                                                                                     0.0s
 => ERROR [deps 4/8] COPY calendso/apps/web/package.json calendso/apps/web/yarn.lock ./apps/web/                                                                          0.0s
 => ERROR [deps 5/8] COPY calendso/packages/prisma/package.json ./packages/prisma/package.json                                                                            0.0s
 => ERROR [deps 6/8] COPY calendso/packages/lib/package.json ./packages/lib/package.json                                                                                  0.0s
 => ERROR [deps 7/8] COPY calendso/packages/tsconfig/package.json ./packages/tsconfig/package.json                                                                        0.0s
------
 > [builder 3/7] COPY calendso/package.json calendso/yarn.lock calendso/turbo.json ./:
------
------
 > [builder 4/7] COPY calendso/apps/web ./apps/web:
------
------
 > [builder 5/7] COPY calendso/packages ./packages:
------
------
 > [runner  4/14] COPY calendso/package.json calendso/yarn.lock calendso/turbo.json ./:
------
------
 > [deps 3/8] COPY calendso/package.json calendso/yarn.lock ./:
------
------
 > [deps 4/8] COPY calendso/apps/web/package.json calendso/apps/web/yarn.lock ./apps/web/:
------
------
 > [deps 5/8] COPY calendso/packages/prisma/package.json ./packages/prisma/package.json:
------
------
 > [deps 6/8] COPY calendso/packages/lib/package.json ./packages/lib/package.json:
------
------
 > [deps 7/8] COPY calendso/packages/tsconfig/package.json ./packages/tsconfig/package.json:
------
failed to compute cache key: "/calendso/yarn.lock" not found: not found
ERROR: Service 'calendso' failed to build : Build failed
```

**Comment by searsaw** [comment]
@leprodude yeah i keep forgetting that there is a submodule to bring in. With your existing repo, run `git submodule update --init --recursive`. That should bring in the `calendso` submodule.

**Comment by leprodude** [comment]
What a great help, thanks to you I was able to successfully build it! Took a while, but now it's done! Thank you so much @searsaw <3

To recap:
```
git clone --recursive git@github.com:calcom/docker.git
cd docker
git remote add searsaw git@github.com:searsaw/docker.git
git fetch --all
git checkout searsaw/fix-workspaces-build-issues
mv .env.example .env #change content as needed
git submodule update --init --recursive
docker-compose up --build
```

I guess these steps are gonna be valid until the pull request gets accepted

**Comment by deeztek** [comment]
Here's how I got it working. Very similar to @leprodude but I didn't do the **git submodule update --init --recursive** command:

```
git clone --recursive https://github.com/calendso/docker.git calendso-docker
cd calendso-docker/
git remote add searsaw https://github.com/searsaw/docker.git
git fetch --all
git checkout searsaw/fix-workspaces-build-issues
mv .env.example .env
docker-compose up --build -d
```

**Comment by BahrouniYousra** [comment]
@searsaw  I check out your branch but I still have problems on installing yarn
Have you any idea how can I fixe it plz ?
------
 > [deps 8/8] RUN yarn install:
#18 3.590 yarn install v1.22.17
#18 3.713 [1/5] Validating package.json...
#18 3.721 [2/5] Resolving packages...
#18 7.342 error Couldn't find package "@calcom/ui@*" required by "@calcom/web@1.0.0" on the "npm" registry.
#18 7.342 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
#18 7.346 Error: Couldn't find package "@calcom/config@*" required by "@calcom/web@1.0.0" on the "npm" registry.
#18 7.346     at MessageError.ExtendableBuiltin (/opt/yarn-v1.22.17/lib/cli.js:721:66)
#18 7.346     at new MessageError (/opt/yarn-v1.22.17/lib/cli.js:750:123)
#18 7.346     at PackageRequest.<anonymous> (/opt/yarn-v1.22.17/lib/cli.js:36582:17)
#18 7.346     at Generator.throw (<anonymous>)
#18 7.346     at step (/opt/yarn-v1.22.17/lib/cli.js:310:30)
#18 7.346     at /opt/yarn-v1.22.17/lib/cli.js:323:13
#18 7.346     at processTicksAndRejections (internal/process/task_queues.js:95:5)

**Comment by saschafoerster** [comment]
I have followed the steps and the same issue as @BahrouniYousra.

**Comment by nickandre10x** [comment]
This worked for me on OSX 12.2

```
gh pr checkout 86
git submodule update --init --recursive
docker-compose up --build
```

Got an error without running the submodule update, not sure if that fixed the issue or the problem is a heisenbug:

```
 > [deps 8/8] RUN yarn install:                                                                                                                                                                             
#17 0.290 yarn install v1.22.17                                                                                                                                                                             
#17 0.375 [1/5] Validating package.json...                                                                                                                                                                  
#17 0.385 [2/5] Resolving packages...                                                                                                                                                                       
#17 1.056 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.                                                                                                          
#17 1.056 error Couldn't find package "@calcom/ui@*" required by "@calcom/web@1.0.0" on the "npm" registry.                                                                                                 
#17 1.058 Error: Couldn't find package "@calcom/config@*" required by "@calcom/web@1.0.0" on the "npm" registry.                                                                                            
#17 1.058     at MessageError.ExtendableBuiltin (/opt/yarn-v1.22.17/lib/cli.js:721:66)                                                                                                                      
#17 1.058     at new MessageError (/opt/yarn-v1.22.17/lib/cli.js:750:123)                                                                                                                                   
#17 1.058     at PackageRequest.<anonymous> (/opt/yarn-v1.22.17/lib/cli.js:36582:17)                                                                                                                        
#17 1.058     at Generator.throw (<anonymous>)
#17 1.058     at step (/opt/yarn-v1.22.17/lib/cli.js:310:30)
#17 1.058     at /opt/yarn-v1.22.17/lib/cli.js:323:13
#17 1.058     at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

**Comment by davidknezic** [comment]
Hey @searsaw, thank you for this PR. It helped me build and self-host cal.com.

I noticed that the current Docker image is 2GB in size. I was able to reduce that by a couple 100MBs by switching all stages from `node:14` to `node:14-alpine`. What's more, with this change it's not necessary to install `netcat` anymore, as `nc` is available by default.

**Review by krumware — APPROVED** [review]
LGTM

**Comment by AmoghSaxena** [comment]
@searsaw Thanks! this PR works fine and it helped me a lot!.

**Comment by krumware** [comment]
Reviewing this on a fresh start locally

**Comment by krumware** [comment]
Does work with the caveat that the submodule init steps may need to be updated/clarified

**Pull request merged**
